### DISCUSSION TITLE
Lesson 7: [FAIL. Reason: Not_Sufficient_Amount()] testFundUpdatesDataStructure()

### DISCUSSION THREAD

**Discussion by deep0072** [Cyfrin/foundry-full-course-cu#173] [Q&A]
when i run the forge test showing error   
```Test result: FAILED. 4 passed; 1 failed; finished in 5.36ms

Failing tests:
Encountered 1 failing test in test/FundMe.t.sol:FundMeTest
[FAIL. Reason: Not_Sufficient_Amount()] testFundUpdatesDataStructure() (gas: 57766)```


here is the code 


fundMe.sol
```// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "./priceConverter.sol";

error Not_Sufficient_Amount();
error UnAuthorised();

contract FundMe {
    using PriceConverter for uint256;

    uint256 public constant MINIMUM_USD = 5e18;

    address[] public s_funders;

    mapping(address => uint256) public s_amountToFunded;

    address public immutable s_owner;
    AggregatorV3Interface private s_priceFeed;
    uint256 ethPriceUsd;

    constructor(address priceFeed) {
        s_owner = msg.sender;
        s_priceFeed = AggregatorV3Interface(priceFeed);
    }

    modifier onlyOwner() {
        if (msg.sender != s_owner) {
            revert UnAuthorised();
        }
        _;
    }

    function getVersion() public view returns (uint256) {
        uint256 version = s_priceFeed.version();

        return version;
    }

    function fund() public payable {
        ethPriceUsd = msg.value.conversionRate(s_priceFeed);
        if (ethPriceUsd < MINIMUM_USD) {
            revert Not_Sufficient_Amount();
        }

        s_funders.push(msg.sender);
        s_amountToFunded[msg.sender] += msg.value;
    }

    function withDraw() public onlyOwner {
        for (
            uint256 funderIndex;
            funderIndex < s_funders.length;
            funderIndex++
        ) {
            address funder = s_funders[funderIndex];
            s_amountToFunded[funder] = 0;
        }

        s_funders = new address[](0);

        (bool success, ) = payable(msg.sender).call{
            value: address(this).balance
        }("");

        require(success, "txn failed");
    }

    function getAddressAmountToFunded(address _funder)
        public
        view
        returns (uint256)
    {
        uint256 amount = s_amountToFunded[_funder];
        return amount;
    }

    function getFunderWithIndex(uint256 funderIndex)
        public
        view
        returns (address)
    {
        return s_funders[funderIndex];
    }

    function getEthPriceUsd() public view returns (uint256) {
        return ethPriceUsd;
    }

    receive() external payable {
        fund();
    }
}```

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import {Test, console} from "forge-std/Test.sol";
import "../src/FundMe.sol";
import {FundMeScript} from "../script/Fundme.s.sol";

contract FundMeTest is Test {
    // public variable is fundMe contract
    FundMe public fundMe;
    // makeAddr function used to create random address

    address User = makeAddr("Deepak"); // User is the funder here
    uint256 constant SEND_VALUE = 1 ether; // amount to fund the contract
    uint256 constant STARTING_BALANCE = 10e18; // amount to fund the contract

    // setUp function runs always first and deploy contract.
    function setUp() external {
        // here this test file is the deployer of the fundMe contract
        // fundMe = new FundMe(address(this));

        // if run this then msg.sender is us not the this test file
        FundMeScript deployFundMe = new FundMeScript();
        fundMe = deployFundMe.run();
    }

    function testMinimumDollarIsFive() public {
        assertEq(fundMe.MINIMUM_USD(), 5e18);
    }

    function testOwnerIsMsgSender() public {
        console.log(fundMe.s_owner());
        console.log(address(this));
        assertEq(fundMe.s_owner(), msg.sender);
    }

    function testPriceFeedversionIsAccurate() public {
        uint256 version = fundMe.getVersion();
        assertEq(version, 4);
    }

    function testFundFailsWithoutEnoughETH() public {
        // this test supposed to be passed when we don't have enough eth to
        vm.expectRevert(); // the next line should revert

        fundMe.fund(); // here we are trying to send the zero eth to
    }

    function testFundUpdatesDataStructure() public {
        vm.deal(User, STARTING_BALANCE); // deal method set address to new balance
        console.log(User.balance, "balance");
        vm.prank(User); // set the msg.sender means the code below will be executed by "User"
        fundMe.fund{value: SEND_VALUE}();
        console.log(address(fundMe).balance, "balance");

        uint256 amount = fundMe.getAddressAmountToFunded(User);
        assertEq(amount, SEND_VALUE);
    }
}```

fundme.s.sol
```// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {Test, console} from "forge-std/Test.sol";

import "../src/FundMe.sol";
import {HelperConfig} from "../script/HelperConfig.s.sol";

contract FundMeScript is Script, Test {
    function run() public returns (FundMe) {
        // before startBroadcasting so that it will not cost any gas
        HelperConfig helperConfig = new HelperConfig();
        address ethUsdPriceFeed = helperConfig.ActivateConfig();
        console.log(ethUsdPriceFeed, "eth price usd");
        vm.startBroadcast();
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        vm.stopBroadcast();
        return fundMe;
    }
}```

helperConfig.sol

```
// SPDX-License-Identifier MIT License

// this script is used to get get price feed address on the basis of chain id

// if chain id on mainnet.then we will get the address of mainnet chainlink priceFeed

pragma solidity ^0.8.18;

import {MockV3Aggregator} from "test/mock/MockV3Aggregator.sol";

import {Script} from "forge-std/Script.sol";

contract HelperConfig is Script {
    uint8 constant DECIMAL = 8;
    int256 constant INTIAL_PRICE = 2e8;

    struct NetworkConfig {
        address priceFeedAddress;
    }

    NetworkConfig public ActivateConfig;

    constructor() {
        if (block.chainid == 1) {
            ActivateConfig = getMainnetConfig();
        } else if (block.chainid == 11155111) {
            ActivateConfig = getSepoliaConfig();
        } else {
            ActivateConfig = getAnvilConfig();
        }
    }

    function getSepoliaConfig() public pure returns (NetworkConfig memory) {
        NetworkConfig memory sepoliaConfig = NetworkConfig({
            priceFeedAddress: 0x694AA1769357215DE4FAC081bf1f309aDC325306
        });

        return sepoliaConfig;
    }

    function getAnvilConfig() public returns (NetworkConfig memory) {
        if (ActivateConfig.priceFeedAddress != address(0)) {
            return ActivateConfig;
        }
        vm.startBroadcast();
        // here we are tryint to get the price feed address on
        //local chain by deploying chainlink aggregator on anvil local chain

        MockV3Aggregator mockPriceFeed = new MockV3Aggregator(
            DECIMAL,
            INTIAL_PRICE
        );

        vm.stopBroadcast();

        NetworkConfig memory anvilConfig = NetworkConfig({
            priceFeedAddress: address(mockPriceFeed)
        });

        return anvilConfig;
    }

    function getMainnetConfig() public pure returns (NetworkConfig memory) {
        NetworkConfig memory mainnetConfig = NetworkConfig({
            priceFeedAddress: 0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419
        });

        return mainnetConfig;
    }
}```

priceConverter.sol

```// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
    function getPrice(AggregatorV3Interface priceFeed) internal view returns (uint256) {
        (, int256 answer,,,) = priceFeed.latestRoundData();

        return uint256(answer * 1e10);
    }

    function conversionRate(uint256 amountOfEth, AggregatorV3Interface priceFeed) internal view returns (uint256) {
        uint256 price = getPrice(priceFeed);

        uint256 ethPriceUsd = (price * amountOfEth) / 1e18;
        return ethPriceUsd;
    }
}```

mockV3Aggregator.sol
```// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

/**
 * @title MockV3Aggregator
 * @notice Based on the FluxAggregator contract
 * @notice Use this contract when you need to test
 * other contract's ability to read data from an
 * aggregator contract, but how the aggregator got
 * its answer is unimportant
 */
contract MockV3Aggregator {
    uint256 public constant version = 4;

    uint8 public decimals;
    int256 public latestAnswer;
    uint256 public latestTimestamp;
    uint256 public latestRound;

    mapping(uint256 => int256) public getAnswer;
    mapping(uint256 => uint256) public getTimestamp;
    mapping(uint256 => uint256) private getStartedAt;

    constructor(uint8 _decimals, int256 _initialAnswer) {
        decimals = _decimals;
        updateAnswer(_initialAnswer);
    }

    function updateAnswer(int256 _answer) public {
        latestAnswer = _answer;
        latestTimestamp = block.timestamp;
        latestRound++;
        getAnswer[latestRound] = _answer;
        getTimestamp[latestRound] = block.timestamp;
        getStartedAt[latestRound] = block.timestamp;
    }

    function updateRoundData(uint80 _roundId, int256 _answer, uint256 _timestamp, uint256 _startedAt) public {
        latestRound = _roundId;
        latestAnswer = _answer;
        latestTimestamp = _timestamp;
        getAnswer[latestRound] = _answer;
        getTimestamp[latestRound] = _timestamp;
        getStartedAt[latestRound] = _startedAt;
    }

    function getRoundData(uint80 _roundId)
        external
        view
        returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)
    {
        return (_roundId, getAnswer[_roundId], getStartedAt[_roundId], getTimestamp[_roundId], _roundId);
    }

    function latestRoundData()
        external
        view
        returns (uint80 roundId, int256 answer, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound)
    {
        return (
            uint80(latestRound),
            getAnswer[latestRound],
            getStartedAt[latestRound],
            getTimestamp[latestRound],
            uint80(latestRound)
        );
    }

    function description() external pure returns (string memory) {
        return "v0.6/tests/MockV3Aggregator.sol";
    }
}```

**Comment by deep0072** [comment]
guys fixed the issue myself. man this is very irritating. just in helperConfig.sol i changed the int256 constant INTIAL_PRICE = 2e8 to int256 constant INTIAL_PRICE = 2000e8; ;

**Reply by SuperHumano** [reply]
U saved man, Thanks!
